src/components/layout/ChatArea.tsx

```diff
@@ -14,10 +14,37 @@ import {
   FaUserPlus,
 } from "react-icons/fa";
 import ircClient from "../../lib/ircClient";
+import { IsUserTyping } from "../../lib/ircUtils";
 import useStore from "../../store";
-import type { Message as MessageType } from "../../types";
+import type { Message as MessageType, User } from "../../types";
 import EmojiSelector from "../ui/EmojiSelector";
 
+const EMPTY_ARRAY: User[] = [];
+
+export const TypingIndicator: React.FC<{
+  serverId: string;
+  channelId: string;
+}> = ({ serverId, channelId }) => {
+  const key = `${serverId}-${channelId}`;
+
+  const typingUsers = useStore(
+    (state) => state.typingUsers[key] ?? EMPTY_ARRAY,
```

Comment:
Should you really use a const array here instead of just `[]` ? Is this passed as a reference or copy? Feels weird.